src/Microsoft.AspNet.OData.Shared/EnableQueryAttribute.cs

```diff
@@ -371,6 +371,11 @@ public int MaxOrderByNodeCount
             }
         }
 
+        /// <summary>
+        /// Gets the query settings.
+        /// </summary>
+        protected ODataQuerySettings QuerySettings => _querySettings;
```

Comment:
Yeah, this whole thing is a mess. I don't feel comfortable signing off as-is because `_querySettings` sort of exists in memory for the lifetime of the process, but it's also very mutable, so it's really easy for a caller to make mistakes. I see two options to address that:

1. Change this code to use the `CopyFrom` method on `ODataQuerySettings`, that way the "original" settings are always preserved
2. Add a new type for `ODataQuerySettings` that is immutable, along with an overload on `ODataQueryOptions` which uses the new type.

Reply:
Thank you for calling this out - I forgot that this is a `FilterAttribute` and as a result it's cached. Given that, you're right that `_querySettings` spans requests and should therefore be protected. I'll look into option 1 for now as I don't think 2 is worth it given how little data `CopyFrom` will copy.

Reply:
Out of curiosity, does this concern also apply to the existing properties on this class which appear to mutate the `_querySettings` when modified (e.g. the `PageSize`)?

Reply:
I believe so - the "normal" use-case is that a developer sticks the attribute on a controller/action and, using attribute syntax, populates the required property values. There's no way to modify those at runtime.

Where thread-safety comes into the picture is if you inherit from `EnableQueryAttribute`, at which point you can theoretically modify the values per-request which, given the semi-static nature of `EnableQueryAttribute` instances, can lead to cross-thread state issues.

So the "bigger picture" questions are how can we allow developers to override some aspects of `EnableQueryAttribute` without introducing potential race conditions.

Reply:
@ificator are you waiting on something here? I think I'm still happy with either of the two options I put at the beginning. I don't see thread-safety issues with either of them (happy to be wrong though).

Reply:
Sorry for the delay! I've changed my approach to avoid the need to expose `ODataQuerySettings` for now (although it probably should be done at some point) by focusing more on my specific scenario. I think this latest change should be good.